librz/core/cmd/cmd_debug.c

```diff
@@ -2235,51 +2199,214 @@ RZ_IPI RzCmdStatus rz_cmd_debug_step_until_optype_handler(RzCore *core, int argc
+	RzDebugTracepoint *t;
+	if ((t = rz_debug_trace_get(core->dbg,
+		     rz_num_math(core->num, argv[1])))) {
+		rz_cons_printf("offset = 0x%" PFMT64x "\n", t->addr);
+		rz_cons_printf("opsize = %d\n", t->size);
+		rz_cons_printf("times = %d\n", t->times);
+		rz_cons_printf("count = %d\n", t->count);
+		// TODO cons_printf("time = %d\n", t->tm);
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+// dtl
+RZ_IPI RzCmdStatus rz_cmd_debug_traces_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	rz_debug_trace_list(core->dbg, mode, core->offset);
+	return RZ_CMD_STATUS_OK;
+}
+
+// dt+
+RZ_IPI RzCmdStatus rz_cmd_debug_trace_add_handler(RzCore *core, int argc, const char **argv) {
+	ut64 addr = rz_num_math(core->num, argv[1]);
+	int count = argc > 2 ? rz_num_math(core->num, argv[2]) : 1;
+	RzAnalysisOp *op = rz_core_op_analysis(core, addr, RZ_ANALYSIS_OP_MASK_HINT);
+	if (op) {
+		RzDebugTracepoint *tp = rz_debug_trace_add(core->dbg, addr, op->size);
+		if (!tp) {
+			rz_analysis_op_free(op);
+			return RZ_CMD_STATUS_OK;
+		}
+		tp->count = count;
+		rz_analysis_trace_bb(core->analysis, addr);
+		rz_analysis_op_free(op);
+	} else {
+		RZ_LOG_ERROR("Cannot analyze opcode at 0x%08" PFMT64x "\n", addr);
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+// dt++
+RZ_IPI RzCmdStatus rz_cmd_debug_trace_add_addrs_handler(RzCore *core, int argc, const char **argv) {
+	for (int i = 1; i < argc; ++i) {
+		ut64 addr = rz_num_get(NULL, argv[i]);
+		(void)rz_debug_trace_add(core->dbg, addr, 1);
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+// dt-
+RZ_IPI RzCmdStatus rz_cmd_debug_traces_reset_handler(RzCore *core, int argc, const char **argv) {
+	rz_tree_reset(core->dbg->tree);
+	rz_debug_trace_free(core->dbg->trace);
+	rz_debug_tracenodes_reset(core->dbg);
+	core->dbg->trace = rz_debug_trace_new();
+	return RZ_CMD_STATUS_OK;
+}
+
+// dta
+RZ_IPI int rz_cmd_debug_trace_addr(void *data, const char *input) {
+	RzCore *core = (RzCore *)data;
+	rz_debug_trace_at(core->dbg, input);
+	return 0;
+}
+
+// dtc
+RZ_IPI int rz_cmd_debug_trace_dtc(void *data, const char *input) {
+	RzCore *core = (RzCore *)data;
+	debug_trace_calls(core, input);
+	return RZ_CMD_STATUS_OK;
+}
+
+// dtd
+RZ_IPI RzCmdStatus rz_cmd_debug_traces_dtd_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	int min = argc > 1 ? (int)rz_num_math(core->num, argv[1]) : 0;
+	RzDebugTracepoint *trace;
+	RzListIter *iter;
+	RzAnalysisOp *op;
+	int n = 0;
+
+	switch (mode) {
+	case RZ_OUTPUT_MODE_QUIET:
+		rz_list_foreach (core->dbg->trace->traces, iter, trace) {
+			if (n >= min) {
+				rz_cons_printf("%d  ", trace->count);
+				rz_cons_printf("0x%08" PFMT64x "\n", trace->addr);
+				break;
+			}
+			n++;
+		}
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		rz_list_foreach (core->dbg->trace->traces, iter, trace) {
+			op = rz_core_analysis_op(core, trace->addr, RZ_ANALYSIS_OP_MASK_BASIC | RZ_ANALYSIS_OP_MASK_DISASM);
```

Comment:
i suggest to put the rz_list_foreach outside of the switch, because no matter what the output mode is you are going to need to iterate over the traces anyway.